mlir/lib/Linker/IRMover.cpp

```diff
@@ -7,7 +7,354 @@
 //===----------------------------------------------------------------------===//
 
 #include "mlir/Linker/IRMover.h"
+#include "llvm/Support/Error.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/Verifier.h"
+
+#include <assert.h>
+#include <optional>
 
 using namespace mlir;
+using llvm::Error;
+
+namespace {
+
+class MLIRLinker {
+  Operation *composite;
+  OwningOpRef<Operation *> src;
+
+  // TODO: This is a ValueToValueMapTy in llvm-link. I'm assuming this would be
+  // the equivalent in mlir.
+  IRMapping valueMap;
+
+  DenseSet<Operation *> valuesToLink;
+  std::vector<Operation *> worklist;
+  // Replace-all-uses-with worklist
+  std::vector<std::pair<Operation *, Operation *>> rauwWorklist;
+
+  void maybeAdd(Operation *val) {
+    if (valuesToLink.insert(val).second)
+      worklist.push_back(val);
+  }
+
+  std::optional<Error> foundError;
+  void setError(Error e) {
+    if (e)
+      foundError = std::move(e);
+  }
+
+  Error linkFunctionBody(Operation *dst, FunctionLinkageOpInterface src);
+  Error linkGlobalValueBody(Operation *dst, Operation *src);
```

Comment:
why is src operation and nto an itnerface when `linkFunctionBody` takes src as interface?

Reply:
good question... Should be unified.